agenthub/codeact_agent/codeact_agent.py

```diff
@@ -194,27 +215,148 @@ def step(self, state: State) -> Action:
+                break
+
+        # TODO: Add functionality for preserving last N messages
+        # MESSAGE_SUMMARY_TRUNC_KEEP_N_LAST = 3
+        # if preserve_last_N_messages:
+        #     candidate_messages_to_summarize = candidate_messages_to_summarize[:-MESSAGE_SUMMARY_TRUNC_KEEP_N_LAST]
+        #     token_counts = token_counts[:-MESSAGE_SUMMARY_TRUNC_KEEP_N_LAST]
+
+        print(f'MESSAGE_SUMMARY_TRUNC_TOKEN_FRAC={MESSAGE_SUMMARY_TRUNC_TOKEN_FRAC}')
+        # print(f'MESSAGE_SUMMARY_TRUNC_KEEP_N_LAST={MESSAGE_SUMMARY_TRUNC_KEEP_N_LAST}')
+        print(f'token_counts={token_counts}')
+        print(f'message_buffer_token_count={message_buffer_token_count}')
+        print(f'desired_token_count_to_summarize={desired_token_count_to_summarize}')
+        print(
+            f'len(candidate_messages_to_summarize)={len(candidate_messages_to_summarize)}'
+        )
+
+        if len(candidate_messages_to_summarize) == 0:
+            raise SummarizeError(
+                f"Summarize error: tried to run summarize, but couldn't find enough messages to compress [len={len(messages)}]"
+            )
+
+        # TODO: Try to make an assistant message come after the cutoff
+
+        message_sequence_to_summarize = candidate_messages_to_summarize
+
+        if len(message_sequence_to_summarize) <= 1:
+            # This prevents a potential infinite loop of summarizing the same message over and over
+            raise SummarizeError(
+                f"Summarize error: tried to run summarize, but couldn't find enough messages to compress [len={len(message_sequence_to_summarize)} <= 1]"
+            )
+        else:
+            print(
+                f'Attempting to summarize with last summarized event id = {last_summarized_event_id}'
+            )
+
+        summary_action = self.memory_condenser.summarize_messages(
+            message_sequence_to_summarize=message_sequence_to_summarize
+        )
+        summary_action.last_summarized_event_id = last_summarized_event_id
+        print(f'Got summary: {summary_action}')
+        history.add_summary(summary_action)
+        print('Added summary to history')
+
+    def search_memory(self, query: str) -> list[str]:
+        raise NotImplementedError('Implement this abstract method')
+
     def _get_messages(self, state: State) -> list[dict[str, str]]:
         messages = [
             {'role': 'system', 'content': self.system_message},
             {'role': 'user', 'content': self.in_context_example},
         ]
 
         for event in state.history.get_events():
+            if isinstance(event, AgentSummarizeAction):
+                action_message = get_action_message(event)
+                if action_message:
+                    messages.append(action_message)
+                observation_message = get_observation_message(event)
+                if observation_message:
+                    messages.append(observation_message)
+                continue
```

Comment:
Do we skip all the action and observations before `AgentSummarizeAction` or no? (i think we should?)

Reply:
Corrected it